File: Core/Entities/StringUtility.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Core.Entities
{
    public class StringUtility
    {
        private static readonly string[] VietnameseSigns = new string[]
{
  "aAeEoOuUiIdDyY",
  "áàạảãâấầậẩẫăắằặẳẵ",
  "ÁÀẠẢÃÂẤẦẬẨẪĂẮẰẶẲẴ",
  "éèẹẻẽêếềệểễ",
  "ÉÈẸẺẼÊẾỀỆỂỄ",
  "óòọỏõôốồộổỗơớờợởỡ",
  "ÓÒỌỎÕÔỐỒỘỔỖƠỚỜỢỞỠ",
  "úùụủũưứừựửữ",
  "ÚÙỤỦŨƯỨỪỰỬỮ",
  "íìịỉĩ",
  "ÍÌỊỈĨ",
  "đ",
  "Đ",
  "ýỳỵỷỹ",
  "ÝỲỴỶỸ"
};

        public static string RemoveSign4VietnameseString(string str)
        {

            //Tiến hành thay thế , lọc bỏ dấu cho chuỗi
            for (int i = 1; i < VietnameseSigns.Length; i++)
            {
                for (int j = 0; j < VietnameseSigns[i].Length; j++)
                    str = str.Replace(VietnameseSigns[i][j], VietnameseSigns[0][i - 1]);
            }

            string result = Regex.Replace(str, @"[^a-zA-Z0-9\s]", "");

            result = Regex.Replace(result, @"\s+", "-");

            result = result.Trim('-');

            result = result.ToLower();
            return result;
        }

        public static string RemoveVietnameseSign(string vietnamString)
        {
            Regex regex = new Regex(@"\p{IsCombiningDiacriticalMarks}+");
            string strFormD = vietnamString.Normalize(System.Text.NormalizationForm.FormD);
            string str = regex.Replace(strFormD, String.Empty).Replace('\u0111', 'd').Replace('\u0110', 'D');
            string result = Regex.Replace(str, @"[^a-zA-Z0-9\s]", "");

            result = Regex.Replace(result, @"\s+", "-");

            result = result.Trim('-');

            result = result.ToLower();
            return result;
        }
    }
}

File: client/src/app/shared/models/StringUtility.ts
export class StringUtility {
  private static readonly VietnameseSigns: string[] = [
    'aAeEoOuUiIdDyY',
    'áàạảãâấầậẩẫăắằặẳẵ',
    'ÁÀẠẢÃÂẤẦẬẨẪĂẮẰẶẲẴ',
    'éèẹẻẽêếềệểễ',
    'ÉÈẸẺẼÊẾỀỆỂỄ',
    'óòọỏõôốồộổỗơớờợởỡ',
    'ÓÒỌỎÕÔỐỒỘỔỖƠỚỜỢỞỠ',
    'úùụủũưứừựửữ',
    'ÚÙỤỦŨƯỨỪỰỬỮ',
    'íìịỉĩ',
    'ÍÌỊỈĨ',
    'đ',
    'Đ',
    'ýỳỵỷỹ',
    'ÝỲỴỶỸ',
  ];

  public static removeSign4VietnameseString(str: string): string {
    for (let i = 1; i < this.VietnameseSigns.length; i++) {
      for (let j = 0; j < this.VietnameseSigns[i].length; j++) {
        str = str.replace(
          new RegExp(this.VietnameseSigns[i][j], 'g'),
          this.VietnameseSigns[0][i - 1]
        );
      }
    }

    let result = str.replace(/[^a-zA-Z0-9\s]/g, '');
    result = result.replace(/\s+/g, '-');
    result = result.trim();
    result = result.toLowerCase();
    return result;
  }

  public static removeVietnameseSign(vietnamString: string): string {
    const strFormD = vietnamString.normalize('NFD');
    const str = strFormD
      .replace(/[\u0300-\u036f]/g, '')
      .replace(/đ/g, 'd')
      .replace(/Đ/g, 'D');
    let result = str.replace(/[^a-zA-Z0-9\s]/g, '');
    result = result.replace(/\s+/g, '-');
    result = result.trim();
    result = result.toLowerCase();
    return result;
  }

  public static getDayOfTour(duration: string) {
    const day = duration.trim();
    if (duration == 'Trong ngày') return 0;
    for (let i = 0; i < day.length; i++) {
      if (isNaN(parseInt(day[i], 10))) {
        return parseInt(day.substring(0, i), 10);
      }
    }
    return -1;
  }
}
